cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/DynamicBasicMapReduceContext.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.batch;
+
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.mapreduce.AbstractMapReduce;
+import co.cask.cdap.api.mapreduce.MapReduceContext;
+import co.cask.cdap.api.mapreduce.MapReduceSpecification;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.data.Namespace;
+import co.cask.cdap.data2.datafabric.DefaultDatasetNamespace;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DynamicDatasetContext;
+import co.cask.cdap.data2.dataset2.NamespacedDatasetFramework;
+import co.cask.tephra.TransactionContext;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+/**
+ * The MapReduce program runtime context for use in the {@link AbstractMapReduce#beforeSubmit(MapReduceContext)} and
+ * {@link AbstractMapReduce#onFinish(boolean, MapReduceContext)} methods.
+ *
+ * <p>
+ * The runtime context allows the retrieval of datasets that were not specified in the application specification,
+ * letting users access datasets at runtime that aren't known about at compile time.
+ * </p>
+ */
+public class DynamicBasicMapReduceContext extends DynamicDatasetContext implements MapReduceContext {
+  private static final Logger LOG = LoggerFactory.getLogger(DynamicBasicMapReduceContext.class);
+  private final BasicMapReduceContext mapReduceContext;
+  private final LoadingCache<Long, Map<String, Dataset>> datasetsCache;
+
+  public DynamicBasicMapReduceContext(BasicMapReduceContext mapReduceContext,
+                                      DatasetFramework datasetFramework,
+                                      TransactionContext transactionContext,
+                                      CConfiguration cConf) {
+    super(transactionContext,
+          new NamespacedDatasetFramework(datasetFramework, new DefaultDatasetNamespace(cConf, Namespace.USER)),
+          mapReduceContext.getProgram().getClassLoader(),
+          null,
+          mapReduceContext.getRuntimeArguments());
+    this.mapReduceContext = mapReduceContext;
+    this.datasetsCache = CacheBuilder.newBuilder()
```

Comment:
This seems like a strange contract that DynamicDatasetContext exposes (using a LoadingCache keyed by thread id instead of a ThreadLocal), but that's outside the scope of this change.